Code/Source/solver/mat_models.cpp

```diff
@@ -188,116 +360,120 @@ void get_pk2cc(const ComMod& com_mod, const CepMod& cep_mod, const dmnType& lDmn
   }
 
   // Electromechanics coupling - active strain
-  auto Fe  = F;
-  auto Fa = mat_id(nsd);
-  auto Fai = Fa;
-
-  if (cep_mod.cem.aStrain) {
-    actv_strain(com_mod, cep_mod, ya, nfd, fl, Fa);
-    Fai = mat_inv(Fa, nsd);
-    Fe = mat_mul(F, Fai);
-  }
-
-  double J = mat_det(Fe, nsd);
+  EigenMatrix<nsd> Fe  = F;
+  EigenMatrix<nsd> Fa = EigenMatrix<nsd>::Identity();
+  EigenMatrix<nsd> Fai = Fa;
+
+  // if (cep_mod.cem.aStrain) {
+  //   actv_strain(com_mod, cep_mod, ya, nfd, fl, Fa);
+  //   Fai = Fa.inverse();
+  //   Fe = F * Fai;
+  // }
+
+  Ja = Fa.determinant();
+  double J = Fe.determinant();
   double J2d = pow(J, (-2.0/nd));
   double J4d = J2d*J2d;
 
-  auto Idm = mat_id(nsd);
-  auto C = mat_mul(transpose(Fe), Fe);
-  Array<double> E(nsd,nsd);
-  for (int i = 0; i < nsd; i++) {
-    for (int j = 0; j < nsd; j++) {
-      E(i,j) = 0.50 * (C(i,j) - Idm(i,j));
+  EigenMatrix<nsd> Idm = EigenMatrix<nsd>::Identity();
+  EigenMatrix<nsd> C = Fe.transpose() * Fe;
+  EigenMatrix<nsd> E = 0.50 * (C - Idm);
+
+  EigenMatrix<nsd> Ci = C.inverse();
+  double trE = E.trace();
+  double Inv1 = J2d * C.trace();
+  double Inv2 = 0.50 * (Inv1*Inv1 - J4d * (C*C).trace());
+
+
+  // Initialize elasticity tensor
+  EigenTensor<nsd> CC;
+  CC.setZero();
+
+  // Add volumetric stress and elasticity tensor if not ustruct and volumetric
+  // penalty parameter is non-zero
+  if (!ustruct) {
+    if (!utils::is_zero(Kp)) {
+      double p  = 0.0;
+      double pl = 0.0;
+      get_svol_p(com_mod, cep_mod, stM, J, p, pl);
+      S += p * J * Ci;
+      CC += -2.0 * p * J * ten_symm_prod_eigen<nsd>(Ci, Ci) + pl * J * ten_dyad_prod_eigen<nsd>(Ci, Ci);
     }
   }
 
-  auto Ci = mat_inv(C, nsd);
-  double trE = mat_trace(E, nsd);
-  double Inv1 = J2d * mat_trace(C,nsd);
-  double Inv2 = 0.50 * (Inv1*Inv1 - J4d * mat_trace(mat_mul(C,C), nsd));
-
-  // Contribution of dilational penalty terms to S and CC
-  double p  = 0.0;
-  double pl = 0.0;
-
-  if (!utils::is_zero(Kp)) {
-    get_svol_p(com_mod, cep_mod, stM, J, p, pl);
-  }
-
-  // Now, compute isochoric and total stress, elasticity tensors
-  //
-  Tensor4<double> CC(nsd,nsd,nsd,nsd);
-
+  // Now, add isochoric and total stress, elasticity tensors
   switch (stM.isoType) {
     case ConstitutiveModelType::stIso_lin: {
+      if (ustruct) {
```

Comment:
This check should be made much sooner than after the simulation has been partitioned and material properties are starting to be computed!

And also it is not clear what the meaning of the `ustruct` variable is unless you read the error message.

Reply:
I feel like this may be a more natural place to have this check. I'm not super familiar with the code that performs the file reading, so I'm not sure where to place this check. When running a ustruct simulation with one of these material models, the simulation fails immediately and the error message is fairly clear.

Reply:
@ktbolt Just pinging this when you get you get a chance. After resolving this conversation I think we can merge.

Reply:
@aabrown100-git The problem is that when running on an HPC cluster your job can wait in a queue for a long time and then immediately fail because the input is not correct, would need something like a single core preprocessing check before running the main job on a larger number of cores. 

We can see if users are having any problems with this sort of thing before implementing a solution. Users creating a job from the SV GUI should be mostly OK. 

A check should be made in the `ConstitutiveModelParameters` class I think, calling a function defined in the `mat_models` namespace. I will open an Issue for this and implement the solution.